msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java

```diff
@@ -1914,4 +2040,225 @@ else if (toolId.equals(DiscussionForumService.MESSAGES_TOOL_ID))
+		  if (element.getUserId().equals(rrepMsg.getCreatedBy())) {
+			  return false;
+		  }
+	  }
+	  return true;
+  }
+
+  private boolean errMaxMessageRespond(PrivateMessage rrepMsg) {
+	  int max = ServerConfigurationService.getInt("msgcntr.no.reply.max.respond", 5);
+	  if (getNumMessageRespond(rrepMsg.getCreatedBy(), rrepMsg.getInReplyTo().getId()) > max) {
+		  return true;
+	  }
+	  return false;
+  }
+  
+  private int getNumMessageRespond(final String userId, final Long messageId) {
+	  log.debug("getNumMessageRespond executing");
+
+	  HibernateCallback<Number> hcb = session -> {
+		  Query q = session.getNamedQuery(QUERY_RESPONSED_COUNT);
+		  q.setParameter("userId", userId, StringType.INSTANCE);
+		  q.setParameter("messageId", messageId, LongType.INSTANCE);
+		  return (Number) q.uniqueResult();
+	  };
+    
+	  return getHibernateTemplate().execute(hcb).intValue();
+  }
+  
+  private String getUserIdByFowardMail(final Long messageId, final String mail) {
+	  log.debug("getUserIdByFowardMail executing");
+
+	  HibernateCallback<String> hcb = session -> {
+		  Query q = session.getNamedQuery(QUERY_USER_ID_BY_FOWARD_MAIL);
+		  q.setParameter("messageId", messageId, LongType.INSTANCE);
+		  q.setParameter("mail", mail, StringType.INSTANCE);
+		  return (String)q.uniqueResult();
+	  };
+
+	  return getHibernateTemplate().execute(hcb);
+  }
+
+  public void processPvtMsgReplySentAction(PrivateMessage currentMessage, PrivateMessage rrepMsg) {
+	  if (rrepMsg != null) {
+		  Map<User, Boolean> recipients = getRecipients(rrepMsg.getRecipients());
+
+		  sendPrivateMessage(rrepMsg, recipients, false);
+
+		  if (!rrepMsg.getDraft()) {
+			  markMessageAsRepliedForUser(currentMessage, rrepMsg.getAuthorId());
+			  String contextId = ((PrivateMessageRecipientImpl)currentMessage.getRecipients().get(0)).getContextId();
+			  markMessageAsReadForUser(currentMessage, contextId, rrepMsg.getAuthorId(), DiscussionForumService.MESSAGES_TOOL_ID);
+			  LRS_Statement statement = null;
+			  if (null != learningResourceStoreService) {
+				  try{
+					  statement = getStatementForUserSentPvtMsg(currentMessage.getTitle(), SAKAI_VERB.responded, currentMessage);
+				  }catch(Exception e){
+					  log.error(e.getMessage(), e);
+				  }
+			  }
+
+			  Event event = eventTrackingService.newEvent(DiscussionForumService.EVENT_MESSAGES_RESPONSE, getEventMessage(currentMessage, DiscussionForumService.MESSAGES_TOOL_ID, rrepMsg.getAuthorId(), ((PrivateMessageRecipientImpl)currentMessage.getRecipients().get(0)).getContextId()), null, true, NotificationService.NOTI_OPTIONAL, statement);
+			  eventTrackingService.post(event);
+		  }
+	  }
+  }
+  
+  private static String encrypt(String value) {
+	  try {
+		  IvParameterSpec iv = new IvParameterSpec(INIT_VECTOR.getBytes(StandardCharsets.UTF_8));
+		  String key = ServerConfigurationService.getString("msgcntr.no.reply.secret", "1111111111111111");
```

Comment:
does this key need to be around longer than a restart of the jvm?

Reply:
In our case it's neccesary because we have 24 nodes right which needs to share this secret and the secret is shared and managed and encrypted by other department. 
I hope this answers your question.

Reply:
If one node goes down and it's restarted that secred should be used again.